src/lib/storage/disk_analyzer.rb

```
@@ -0,0 +1,368 @@
+        if candidate_disks.empty?
+          log.info("No candidate disks left after eliminating installation disks")
+          log.info("Trying with non-USB installation disks")
+          candidate_disks = @installation_disks.select { |disk| disk.transport != ::Storage::USB }
+        end
+        if candidate_disks.empty?
+          log.info("Still no candidate disks left")
+          log.info("Trying with installation disks out of sheer desperation")
+          candidate_disks = @installation_disks
+        end
+        candidate_disks
+      end
+
+      # Remove any installation disks from 'disks' and return a disks array
+      # containing the disks that are not installation media.
+      #
+      # @param disks [Array<::Storage::Disk>]
+      # @return [Array<::Storage::Disk>] non-installation disks
+      #
+      def remove_installation_disks(disks)
+        # We can't simply use
+        #   disks -= @installation_disks
+        # because the list elements (from libstorage) don't provide a .hash method.
+        # Comparing device names ("/dev/sda"...) instead.
+
+        inst_names = dev_names(@installation_disks)
+        disks.delete_if { |disk| inst_names.include?(disk.name) }
+      end
+
+      # Find any Linux partitions on any of the candidate disks.
+      # This may be a normal Linux partition (type 0x83), a Linux swap
+      # partition (type 0x82), an LVM partition, or a RAID partition.
+      #
+      # @return [Array<Partition>] Linux partitions
+      #
+      def find_linux_partitions
+        linux_partitions = []
+        @candidate_disks.each do |disk|
+          begin
+            disk.partition_table.partitions.each do |partition|
+              if LINUX_PARTITION_IDS.include?(partition.id)
+                log.info("Found Linux partition #{partition} (ID 0x#{partition.id.to_s(16)})")
+                linux_partitions << partition
+              end
+            end
+          rescue RuntimeError => ex
```

For me the biggest problem with it, is that you silently return only partial result and API user do not see it, which is wrong from my POV. E.g. consider I would like to do some decision based on linux_partitions, if it is not complete, then my decision will be wrong.